Lists and Keys page translation (#7)
reactjs/pt-BR.reactjs.org: content/docs/lists-and-keys.md

---
id: lists-and-keys
title: Listas e Chaves
permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---

Primeiro, vamos rever como transformamos listas em JavaScript.

> Dado o código abaixo, nós usamos a função [`map()`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map) para receber um array de `números` e dobrar o valor de cada um deles. Atribuímos o novo array retornado pela função `map()` para a variável `doubled` e a imprime no console:

Reviewer: ```suggestion
Dado o código abaixo, nós usamos a função [`map()`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map) para receber um array de `números` e dobrar o valor de cada um deles. Atribuímos o novo array retornado pela função `map()` para a variável `doubled` e imprime no console:
```

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

Esse código imprime `[2, 4, 6, 8, 10]` no console.

> Em React, transformar arrays em listas de [elementos](/docs/rendering-elements.html) é praticamente idêntico a isso.

Reviewer 2: ```suggestion
No React, transformar arrays em listas de [elementos](/docs/rendering-elements.html) é praticamente idêntico a isso.
```

);
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/GjPyQr?editors=0011)

Esse código mostra uma lista não ordenada de números entre 1 e 5.

> 

Reviewer 3: Maybe it'd fit better with `Componente de Lista Básico`

> 

Reviewer: ```suggestion 
### Componente de Lista Básico {#basic-list-component}
```

);
```

> Ao executar esse código, você receberá um aviso que uma chave deve ser definida para os itens da lista. `key` é um atributo string especial que você precisa definir ao criar listas de elementos. Iremos analisar os motivos porque isso é importante na próxima seção.

Reviewer 4: ```suggestion
Ao executar esse código, você receberá um aviso que uma chave deve ser definida para os itens da lista. `key` é um atributo string especial que você precisa definir ao criar listas de elementos. Iremos analisar os motivos pelos quais isso é importante na próxima seção.
```

> Ao executar esse código, você receberá um aviso que uma chave deve ser definida para os itens da lista. `key` é um atributo string especial que você precisa definir ao criar listas de elementos. Iremos analisar os motivos pelo qual isso é importante na próxima seção.

Reviewer: ```suggestion
Ao executar esse código, você receberá um aviso que uma chave deve ser definida para os itens da lista. `key` é um atributo string especial que você precisa definir ao criar listas de elementos. Iremos analisar os motivos porque isso é importante na próxima seção.
```

);
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/jrXYRR?editors=0011)

> ## Chaves {#keys}

Reviewer 3: I think it's good to stick with `Keys` as it actually means a named attribute (`key`)

> As chaves ajudam a React identificar quais itens sofreram alterações, adicionadas ou removidas. As chaves devem ser atribuídas aos elementos dentro do array para dar uma identidade estável aos elementos:

Reviewer 4: ```suggestion
As chaves ajudam React a identificar quais itens sofreram alterações, adicionadas ou removidas. As chaves devem ser atribuídas aos elementos dentro do array para dar uma identidade estável aos elementos:
```

> As chaves ajudam a React identificar quais itens sofreram alterações, adicionadas ou removidas. As chaves devem ser atribuídas aos elementos dentro do array para dar uma identidade estável aos elementos:

Reviewer: ```suggestion
As Chaves ajudam React a identificar quais itens sofreram alterações, adicionadas ou removidas. As Chaves devem ser atribuídas aos elementos dentro do array para dar uma identidade estável aos elementos:
```

Reviewer 4: I think in this case we should keep Chaves in lowercase, they only function as a common noun -- they don't have a special meaning in this context.

> As chaves ajudam a React identificar quais itens sofreram alterações, adicionadas ou removidas. As chaves devem ser atribuídas aos elementos dentro do array para dar uma identidade estável aos elementos:

Reviewer 2: ```suggestion
As chaves ajudam a React identificar quais itens sofreram alterações, foram adicionadas ou removidas. As chaves devem ser atribuídas aos elementos dentro do array para dar uma identidade estável aos elementos:
```

);
```

Quando você não possui nenhum ID estável para os itens renderizados, você pode usar o índice do item como chave em último recurso:

```js{2,3}
const todoItems = todos.map((todo, index) =>
  // Apenas faça isso caso os itens não possuam IDs estáveis
  <li key={index}>
    {todo.text}
  </li>
);
```

> Não recomendamos o uso de índices para chave se a ordem dos itens pode ser alterada. Isso pode impactar de forma negativa o desempenho e poderá causar problemas com o estado do componente. Leia o artigo escrito por Robin Pokorny para [uma explicação aprofundada nos impactos negativos de se usar um índice como chave](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Se você não atribuir uma chave de forma explícita para os itens de uma lista, então a React irá utilizar os índices como chave por padrão.

Reviewer 4: ```suggestion
Não recomendamos o uso de índices para chave se a ordem dos itens pode ser alterada. Isso pode impactar de forma negativa o desempenho e poderá causar problemas com o estado do componente. Leia o artigo escrito por Robin Pokorny para [uma explicação aprofundada nos impactos negativos de se usar um índice como chave](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Se você não atribuir uma chave de forma explícita para os itens de uma lista, então React irá utilizar os índices como chave por padrão.
```

> Não recomendamos o uso de índices para chave se a ordem dos itens pode ser alterada. Isso pode impactar de forma negativa o desempenho e poderá causar problemas com o estado do componente. Leia o artigo escrito por Robin Pokorny para [uma explicação aprofundada nos impactos negativos de se usar um índice como chave](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Se você não atribuir uma chave de forma explícita para os itens de uma lista, então a React irá utilizar os índices como chave por padrão.

Reviewer 2: ```suggestion
Não recomendamos o uso de índices para chave se a ordem dos itens pode ser alterada. Isso pode impactar de forma negativa o desempenho e poderá causar problemas com o estado do componente. Leia o artigo escrito por Robin Pokorny para [uma explicação aprofundada nos impactos negativos de se usar um índice como chave](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Se você não atribuir uma chave de forma explícita para os itens de uma lista, então o React irá utilizar os índices como chave por padrão.
```

> Aqui você poderá ver [uma explicação aprofundada sobre o porquê as chaves são necessárias](/docs/reconciliation.html#recursing-on-children) caso você esteja interessado em aprender mais sobre.

Reviewer 4: ```suggestion
Aqui você poderá ver [uma explicação aprofundada sobre o porquê as chaves são necessárias](/docs/reconciliation.html#recursing-on-children) caso você esteja interessado em aprender mais sobre isso.
```

> Aqui você poderá ver [uma explicação aprofundada sobre o porquê as chaves são necessárias](/docs/reconciliation.html#recursing-on-children) caso você esteja interessado em aprender mais sobre.

Reviewer: ```suggestion
Aqui você poderá ver [uma explicação aprofundada sobre o porquê do uso das chaves são necessárias](/docs/reconciliation.html#recursing-on-children) caso você esteja interessado em aprender mais sobre.
```

> Aqui você poderá ver [uma explicação aprofundada sobre o porquê as chaves são necessárias](/docs/reconciliation.html#recursing-on-children) caso você esteja interessado em aprender mais sobre.

Reviewer 2: ```suggestion
Aqui você poderá ver [uma explicação aprofundada sobre o porquê do uso das chaves é necessário](/docs/reconciliation.html#recursing-on-children) caso você esteja interessado em aprender mais sobre.
```

> ### Extrair Componentes com Chaves {#extracting-components-with-keys}

Reviewer: ```suggestion
### Extrair Componentes com Chaves {#extrair-componentes-com-chaves}
```

Reviewer: cc: @cezaraugusto deve traduzir os links dos títulos também não é?

Reviewer 4: Estes links não podem ser traduzidos

Reviewer 5: @fjoshuajr sabe o motivo? parece válido traduzir visto que é uma âncora apenas

Reviewer: > @glaucia86 Os links que aparecem junto aos títulos não podem ser traduzidos.

Então terá que sinalizar aos demais.... teve gente que acabou traduzindo.

Reviewer 6: Qual o veredicto sobre os links? Revisei um PR inteiro que estava com eles traduzidos. Devo voltar lá e apontar todas as mudanças?

Reviewer 4: Foi um erro da minha parte. Desculpem por qualquer impedimento.

Vamos traduzir os links. 😅

);
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)

> Uma boa regra de definição é que os elementos dentro de `map()` devem especificar chaves.

Reviewer 4: Maybe we can use 'regra de ouro' for 'rule of thumb'

Reviewer 5: "via de regra" seems more appropriate. what do you think?

Reviewer: also too.

Author: So, can I change to this?

```
Por via de regra, os elementos dentro de uma função `map()` devem especificar chaves.
```

Reviewer 4: Yes

> Uma boa regra de definição é que os elementos dentro de `map()` devem especificar chaves.

Reviewer 2: ```suggestion
Uma boa regra de ouro é que os elementos dentro de `map()` devem especificar chaves.
```

> ### Chaves devem ser Únicas apenas entre Elementos Irmãos {#keys-must-only-be-unique-among-siblings}

Reviewer: ### Chaves devem ser Únicas apenas entre Elementos Irmãos {#chaves-devem-ser-unicas-apenas-entre-elementos-irmaos}

Reviewer 4: Aqui também não podem ser traduzidos

> Chaves usadas nos arrays devem ser únicas entre seus items irmãos. Contudo elas não precisam ser únicas globalmente. Podemos usar as mesmas chaves ao criar dois arrays diferentes:

Reviewer 4: I think here we should keep siblings as 'elementos irmãos' just like the title

}

const posts = [

>   {id: 1, title: 'Olá Mundo', content: 'Bem vindo para aprender React!'},

Reviewer: ```suggestion
  {id: 1, title: 'Olá Mundo', content: 'Bem-vindo em aprender React!'},
```

>   {id: 2, title: 'Instalação', content: 'Você pode instalar a React utilizando o npm.'}

Reviewer: ```suggestion
  {id: 2, title: 'Instalação', content: 'Você pode instalar o React utilizando o npm.'}
```

Author: It's something that we need to think about because React is a library = biblioteca, and using `o` for me seems odd, what do you think about that?

Reviewer: Responderei em português: 98% dos artigos aqui no Brasil e vídeos tutoriais falam: o React.
Exemplo: https://tableless.com.br/react-o-que-e-e-como-funciona-essa-ferramenta/

Reviewer: What do you think about it @cezaraugusto ? For me I prefer `o React`

>   {id: 2, title: 'Instalação', content: 'Você pode instalar a React utilizando o npm.'}

Reviewer 4: ```suggestion
  {id: 2, title: 'Instalação', content: 'Você pode instalar React utilizando o npm.'}
```

];
ReactDOM.render(
  <Blog posts={posts} />,
  document.getElementById('root')
);
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/NRZYGN?editors=0010)

> As chaves servem como uma dica para a React, mas elas não são passadas para os componentes. Se você precisar do mesmo valor em um componente, defina ele explicitamente como uma `prop` com um nome diferente:

Reviewer: ```suggestion
As chaves servem como uma dica para o React. Mas, elas não são passadas para os componentes. Se você precisar do mesmo valor em um componente, defina ele explicitamente como uma `prop` com um nome diferente:
```

> As chaves servem como uma dica para a React, mas elas não são passadas para os componentes. Se você precisar do mesmo valor em um componente, defina ele explicitamente como uma `prop` com um nome diferente:

Reviewer 4: ```suggestion
As chaves servem como uma dica para React, mas elas não são passadas para os componentes. Se você precisar do mesmo valor em um componente, defina ele explicitamente como uma `prop` com um nome diferente:
```

> As chaves servem como uma dica para a React, mas elas não são passadas para os componentes. Se você precisar do mesmo valor em um componente, defina ele explicitamente como uma `prop` com um nome diferente:

Reviewer 2: ```suggestion
As chaves servem como uma dica para o React, mas elas não são passadas para os componentes. Se você precisar do mesmo valor em um componente, defina ele explicitamente como uma `prop` com um nome diferente:
```

);
```

> No exemplo acima, o componente `Post` pode acessar `props.id`, mas não pode acessar `props.key`.

Reviewer: ```suggestion
No exemplo acima, o componente `Post` pode acessar `props.id`. Mas, não pode acessar `props.key`.
```

> ### Incluindo map() no JSX {#embedding-map-in-jsx}

Reviewer: ```suggestion
### Incluindo map() no JSX {#incluindo-map-no-jsx}
```

Reviewer 4: Aqui também não podem ser traduzidos

Reviewer 4: Resolvido

}
```

[**Experimente no CodePen**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)

> Às vezes isso resulta em um código mais limpo, mas esse padrão também pode ser abusado. Como em JavaScript, depende de você decidir se vale a pena extrair uma variável para aumentar a legibilidade. Lembre-se que se o corpo da função `map()` tiver muitos níveis, poderá ser uma boa hora para [extrair um componente](/docs/components-and-props.html#extracting-components).

Reviewer 4: ```suggestion
Às vezes isso resulta em um código mais limpo, mas esse padrão também pode ser abusado. Como em JavaScript, depende de você decidir se vale a pena extrair uma variável para aumentar a legibilidade. Lembre-se que se o corpo da função `map()` tiver muitos níveis, poderá ser um bom momento para [extrair um componente](/docs/components-and-props.html#extracting-components).
```

> Às vezes isso resulta em um código mais limpo, mas esse padrão também pode ser abusado. Como em JavaScript, depende de você decidir se vale a pena extrair uma variável para aumentar a legibilidade. Lembre-se que se o corpo da função `map()` tiver muitos níveis, poderá ser uma boa hora para [extrair um componente](/docs/components-and-props.html#extracting-components).

Reviewer: ```suggestion
Às vezes isso resulta em um código mais limpo. Mas, esse padrão também pode ser enganado. Como em JavaScript, depende de você decidir se vale a pena extrair uma variável para aumentar a legibilidade. Lembre-se que se o corpo da função `map()` tiver muitos níveis, poderá ser uma boa hora para [extrair um componente](/docs/components-and-props.html#extracting-components).
```

Reviewer: @WendellAdriel abused = abuso
Sinônimo de abuso mais adequado ao contexto em questão: enganar (confundir) ;)

Reviewer 2: Eu acho que é ao invés de enganar a palavra deveria se: `confuso`

Author: My mistake, sorry!!! xD
